src/data/list/basic.lean

```diff
@@ -44,10 +44,6 @@ assume l₁ l₂, assume Pe, tail_eq_of_cons_eq Pe
 
 /- mem -/
 
-theorem eq_nil_of_forall_not_mem : ∀ {l : list α}, (∀ a, a ∉ l) → l = nil
-| []        := assume h, rfl
-| (b :: l') := assume h, absurd (mem_cons_self b l') (h b)
-
 theorem mem_singleton_self (a : α) : a ∈ [a] := mem_cons_self _ _
```

Comment:
I have never understood the conventions here. I'm pretty sure there are other places where we have both P <-> Q and P -> Q, although admittedly the examples that immediately come to mind are in core. If P <-> Q then is proving P -> Q generally discouraged?

Reply:
yes, although there are a variety of exceptions to this, for example if the proof is large and you want to break it up, or if it is useful as a rewrite or conditional simp lemma. I'm on the fence for this particular example, but given that the theorem is not used very much and it's only an extra character or two to refer to the right theorem I'm inclined to accept this deletion.

Reply:
@kbuzzard There are plenty of examples where a `P → Q` proof is used as a component of the proof for `P ↔ Q`, and those are often due to the size of the proof. I think those are perfectly fine.

If `eq_nil_of_forall_not_mem` were used in `eq_nil_iff_forall_not_mem`, then it should probably stay. In this case, however, there is no connection between the proofs, the former is hardly used, and its uses can be easily replaced by the latter. This is why I called `eq_nil_of_forall_not_mem` redundant, since it doesn't appear to be much more useful than `eq_nil_iff_forall_not_mem`.